Add basic search on entity names

senior opened the pull request:
This PR has basic support on the backend for search of entities. This is currently a first cut / work in progress.

TODO
- [x] Search over entities
  - [x] Cards
  - [x] Collections
  - [x] Dashboards
  - [x] Pulses
  - [x] Metrics
  - [x] Segments
- [x] Filtering by
  - [x] Archived?
  - [x] Collection
- [x] Permissions
- [x] Evaluate Card Performance

senior commented on src/metabase/api/search.clj:
```diff
@@ -0,0 +1,49 @@
+  {:select [:name :description :id [(hx/literal entity-name) :type]]
+   :from   [entity]
+   :where  [:and
+            [:like :name search-string]
```
You're right, I was thinking `like` was case-insensitive, but I was confusing it for `ilike` in postgres

salsakran commented on src/metabase/api/search.clj:
```diff
@@ -0,0 +1,49 @@
+            [schema.core :as s]
+            [toucan.db :as db]))
+
+(defn- create-search-query [entity entity-name search-string archived?]
```
This is a little too cute, especially since we know we'll need to extend this for eg tables which have different names (eg "display_name"). Imo it's not worth saving a couple of lines of code in a couple places to have such a rigid expectation of which fields each type's search will hit.

camsaul replied:
could always switch it to be a multimethod that dispatches off of entity down the road.

salsakran replied:
so long as easily predicable things aren't made difficult -
just to write these out, things that are easily predictable:
1. we'll want to search more other model types
2. we'll want to experiment what is shown in the search listing for each of these types
3. we'll have slightly different ranking weights by type (explicitly or implicitly)

senior replied:
FWIW, I think the queries will drift far enough apart soon that I won't be able to get away with what's there, just seemed wrong to start out with a bunch of copy and pasted code when it wasn't needed. In addition to the things you've described, I also think we'll need/want to use some specific query performance hacks for questions. I'm not worrying about it now as we're getting it working, but later versions of MySQL (5.6+) and PG have indexing and functions that would make these kinds of queries faster.

I expect it to be minor refactor to break it up and get more flexibility, just waiting to cross that threshold.

salsakran replied:
cool. my main point is: "make the things that are easily predictable easy"

senior commented:
Related to https://github.com/metabase/metabase/pull/7649

senior commented:
I was a little concerned about performance of the search query with a larger number of cards. I don't think it will be an issue. I just loaded up 3000 cards and ran the `like` query against it, it was around 20ms on PostgreSQL. There are better indexes for dealing with this kind of query, but it doesn't look to be necessary anytime soon.

iethree committed:
Add a search endpoint for quering names across entity types

This adds a new search route that provides a case insensitive query
across the name fields of cards, collections, dashboards, pulses,
segments and metrics. Users will only get search results for the
collections they are able to access.

camsaul reviewed (approved):
This looks fine, except don't we want to add indexes on `lower(name)`? Otherwise we're doing full table scans
